target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_arcadyan_vgv952cjw33-e-ir.dtsi

```diff
@@ -0,0 +1,409 @@
+
+/ {
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x8000000>;
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <100>;
+		
+		rfkill {
+			label = "wps";
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+		
+		reset {
+			label = "reset";
+			gpios = <&gpio 40 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	usb_vbus: regulator-usb-vbus {
+		compatible = "regulator-fixed";
+		regulator-name = "USB_VBUS";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpio = <&gpio 33 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		power_green: power {
+			label = "VGV952CJW33:red:power";
+			gpios = <&gpio 31 GPIO_ACTIVE_HIGH>;
+			default-state = "keep";
+		};
+	};
+
+	spi-gpio {
+		compatible = "spi-gpio";
+		address-cells = <1>;
+		size-cells = <0>;
+		gpio-sck = <&gpio 29 GPIO_ACTIVE_HIGH >;
+		gpio-mosi = <&gpio 30 GPIO_ACTIVE_HIGH >;
+		num-chipselects = <1>;
+		cs-gpios = <&gpio 39 GPIO_ACTIVE_HIGH >;
+
+		hc595: gpio_spi@0 {
+			compatible = "fairchild,74hc595";
+			reg = <0>;
+			registers-number = <1>;
+			spi-max-frequency = <1000000>;
+			spi-cpol = <0>;
+			spi-cpha = <0>;
+			gpio-controller;
+			#gpio-cells = <2>;
+		};
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		out_0 { /*Unknown*/
+			gpio-export,name = "hc595_0";
+			gpio-export,output = <1>;
+			gpios = <&hc595 0 GPIO_ACTIVE_HIGH>;
+		};
+		
+		out_1 { /*Unknown*/
+			gpio-export,name = "hc595_1";
+			gpio-export,output = <1>;
+			gpios = <&hc595 1 GPIO_ACTIVE_HIGH>;
+		};
+
+		/* out_2 is used to reset touch IC and owned by eb904_keypad */
+
+		out_3 { /*Unknown*/
+			gpio-export,name = "hc595_3";
+			gpio-export,output = <1>;
+			gpios = <&hc595 3 GPIO_ACTIVE_HIGH>;
+		};
+		
+		out_dsl_eth { /* Switches DSL line + 100MiB eth or 1GiB ethernet on DSL/WAN input. Relay 1.*/
+			gpio-export,name = "dsl_en";
+			gpio-export,output = <1>;
+			gpios = <&hc595 4 GPIO_ACTIVE_LOW>;
+		};
+		
+		out_5 { /* Switches filter for DSL line On/Off. Relays 2 and 3 simultaneously.*/
+			gpio-export,name = "annexj";
+			gpio-export,output = <1>;
+			gpios = <&hc595 5 GPIO_ACTIVE_HIGH>;
+		};
+		
+		out_6 { /* Switch DSL line betwean ISDN modem and Si3050. Relay 5.*/
+			gpio-export,name = "annexb";
+			gpio-export,output = <1>;
+			gpios = <&hc595 6 GPIO_ACTIVE_HIGH>;
+		};
+		
+		out_7 { /* Switch phone outputs N and F betwean analog DSL line and XS1 output from Lantiq SLIC. Relay 4.*/
+			gpio-export,name = "annexa";
+			gpio-export,output = <1>;
+			gpios = <&hc595 7 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	i2c {
+		compatible = "i2c-gpio";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		gpios = <&gpio 19 GPIO_ACTIVE_HIGH /* sda */
+			&gpio 14 GPIO_ACTIVE_HIGH /* scl */
+		>;
+
+		i2c-gpio,delay-us = <5>;
+		
+		tp: eb904tp@0x14 { /* Add touch panel Support */
+			compatible = "lantiq,eb904_keypad";
```

Comment:
@Quallenauge did you try modeling this is as gpio-keys-polled or somthing that maybe does not need a driver?

i see https://github.com/OpenWrt-Repository/openwrt-feeds-easybox904-driver/blob/master/kernel/touchpad/src/eb904_keypad.c

that there is also LED stuff going on, but maybe those can be modeled as gpio-leds and any interaction could be done with triggers maybe from userland

anyhow this i would not count in as "minimum viable" so something for later and anything related should probably be taken out of the first dts

Reply:
This keypad is connected via I2C, it's a bit more complex than just a bunch of keys connected to one GPIO each. A dedicated driver IS needed, for sure.